File: examples/handler.py
#!/usr/bin/env python3
"""
Python HTTP Request Handler
Usage: python handler.py <json_request_data>

The request data is a JSON object containing:
- method: HTTP method (GET, POST, etc.)
- path: URL path
- query: Query parameters dict
- headers: Request headers dict
- body: Request body string
- contentType: Content-Type header value
- remoteIp: Client IP address
- url: Full request URL
- receivedAt: Timestamp
"""

import sys
import json
from datetime import datetime

def main():
    args = sys.argv[1:]
    json_request_data = args[0] if len(args) > 0 else None

    print('=================================')
    print('  Python HTTP Request Handler')
    print('=================================')
    print()

    if not json_request_data:
        print('No request data provided')
        print('=================================')
        sys.exit(1)

    try:
        # Parse the request data (ExpressJS-like format)
        req = json.loads(json_request_data)

        print('=== REQUEST INFORMATION ===')
        print(f"Method: {req.get('method', 'N/A')}")
        print(f"Path: {req.get('path', 'N/A')}")
        print(f"URL: {req.get('url', 'N/A')}")
        print(f"Remote IP: {req.get('remoteIp', 'N/A')}")
        print(f"Content-Type: {req.get('contentType', 'N/A')}")
        print(f"Received At: {req.get('receivedAt', 'N/A')}")
        print()

        # Query parameters (ExpressJS req.query)
        query = req.get('query', {})
        if query:
            print('=== QUERY PARAMETERS ===')
            for key, value in query.items():
                print(f'  {key}: {value}')
            print()

        # Headers (ExpressJS req.headers)
        headers = req.get('headers', {})
        if headers:
            print('=== REQUEST HEADERS ===')
            for key, value in headers.items():
                print(f'  {key}: {value}')
            print()

        # Body (ExpressJS req.body)
        body = req.get('body', '')
        if body:
            print('=== REQUEST BODY ===')
            print(f'Raw: {body}')
            print()

            # Try to parse as JSON if Content-Type suggests it
            content_type = req.get('contentType', '')
            if 'application/json' in content_type:
                try:
                    body_data = json.loads(body)
                    print('Parsed JSON:')
                    print(json.dumps(body_data, indent=2))
                    print()
                except json.JSONDecodeError:
                    print('(Body is not valid JSON)')
                    print()
        else:
            print('=== REQUEST BODY ===')
            print('(No body)')
            print()

        print('=================================')

        # Example: ExpressJS-like usage
        # if req.get('method') == 'POST' and req.get('path') == '/deploy':
        #     body_data = json.loads(req.get('body', '{}'))
        #     if body_data.get('action') == 'deploy':
        #         print('Deploying...')
        #
        # if req.get('query', {}).get('action') == 'start':
        #     print('Starting service...')

    except json.JSONDecodeError as e:
        print(f'Error parsing request data: {e}')
        print(f'Raw input: {json_request_data}')
        print('=================================')
        sys.exit(1)

if __name__ == '__main__':
    main()

File: examples/handler.js
#!/usr/bin/env node

// Node.js HTTP Request Handler
// Usage: node handler.js <json_request_data>
//
// The request data is a JSON object containing:
// - method: HTTP method (GET, POST, etc.)
// - path: URL path
// - query: Query parameters object
// - headers: Request headers object
// - body: Request body string
// - contentType: Content-Type header value
// - remoteIp: Client IP address
// - url: Full request URL
// - receivedAt: Timestamp

const args = process.argv.slice(2);
const jsonRequestData = args[0];

console.log('=================================');
console.log('  Node.js HTTP Request Handler');
console.log('=================================');
console.log();

if (!jsonRequestData) {
    console.log('No request data provided');
    console.log('=================================');
    process.exit(1);
}

// Check if we received the old format (just a path string) vs new format (JSON)
if (jsonRequestData.startsWith('/') && !jsonRequestData.startsWith('{')) {
    console.error('ERROR: Received old format (path string) instead of JSON.');
    console.error('This usually means the Runner app needs to be rebuilt.');
    console.error('Received:', jsonRequestData);
    console.error('');
    console.error('Expected format: JSON object with method, path, query, headers, body, etc.');
    console.log('=================================');
    process.exit(1);
}

try {
    // Parse the request data (ExpressJS-like format)
    const req = JSON.parse(jsonRequestData);

    console.log('=== REQUEST INFORMATION ===');
    console.log(`Method: ${req.method || 'N/A'}`);
    console.log(`Path: ${req.path || 'N/A'}`);
    console.log(`URL: ${req.url || 'N/A'}`);
    console.log(`Remote IP: ${req.remoteIp || 'N/A'}`);
    console.log(`Content-Type: ${req.contentType || 'N/A'}`);
    console.log(`Received At: ${req.receivedAt || 'N/A'}`);
    console.log();

    // Query parameters (ExpressJS req.query)
    if (req.query && Object.keys(req.query).length > 0) {
        console.log('=== QUERY PARAMETERS ===');
        for (const [key, value] of Object.entries(req.query)) {
            console.log(`  ${key}: ${value}`);
        }
        console.log();
    }

    // Headers (ExpressJS req.headers)
    if (req.headers && Object.keys(req.headers).length > 0) {
        console.log('=== REQUEST HEADERS ===');
        for (const [key, value] of Object.entries(req.headers)) {
            console.log(`  ${key}: ${value}`);
        }
        console.log();
    }

    // Body (ExpressJS req.body)
    if (req.body) {
        console.log('=== REQUEST BODY ===');
        console.log('Raw:', req.body);
        console.log();

        // Try to parse as JSON if Content-Type suggests it
        if (req.contentType && req.contentType.includes('application/json')) {
            try {
                const bodyData = JSON.parse(req.body);
                console.log('Parsed JSON:');
                console.log(JSON.stringify(bodyData, null, 2));
                console.log();
            } catch (e) {
                console.log('(Body is not valid JSON)');
                console.log();
            }
        }
    } else {
        console.log('=== REQUEST BODY ===');
        console.log('(No body)');
        console.log();
    }

    console.log('=================================');

    // Example: ExpressJS-like usage
    // if (req.method === 'POST' && req.path === '/deploy') {
    //     const body = JSON.parse(req.body);
    //     if (body.action === 'deploy') {
    //         console.log('Deploying...');
    //     }
    // }
    //
    // if (req.query.action === 'start') {
    //     console.log('Starting service...');
    // }

} catch (error) {
    console.error('Error parsing request data:', error.message);
    console.error('Raw input:', jsonRequestData);
    console.log('=================================');
    process.exit(1);
}
